core/src/main/java/org/elasticsearch/action/get/TransportMultiGetAction.java

```diff
@@ -64,26 +64,21 @@ protected void doExecute(final MultiGetRequest request, final ActionListener<Mul
         for (int i = 0; i < request.items.size(); i++) {
             MultiGetRequest.Item item = request.items.get(i);
 
-            if (!clusterState.metaData().hasConcreteIndex(item.index())) {
-                responses.set(i, newItemFailure(item.index(), item.type(), item.id(), new IndexNotFoundException(item.index())));
-                continue;
-            }
-
             String concreteSingleIndex;
             try {
-                item.routing(clusterState.metaData().resolveIndexRouting(item.parent(), item.routing(), item.index()));
                 concreteSingleIndex = indexNameExpressionResolver.concreteSingleIndex(clusterState, item).getName();
-
-                if ((item.routing() == null) && (clusterState.getMetaData().routingRequired(concreteSingleIndex, item.type()))) {
-                    String message = "routing is required for [" + concreteSingleIndex + "]/[" + item.type() + "]/[" + item.id() + "]";
-                    responses.set(i, newItemFailure(concreteSingleIndex, item.type(), item.id(), new IllegalArgumentException(message)));
-                    continue;
-                }
             } catch (Exception e) {
                 responses.set(i, newItemFailure(item.index(), item.type(), item.id(), e));
                 continue;
             }
 
+            item.routing(clusterState.metaData().resolveIndexRouting(item.parent(), item.routing(), concreteSingleIndex));
+            if ((item.routing() == null) && (clusterState.getMetaData().routingRequired(concreteSingleIndex, item.type()))) {
+                String message = "routing is required for [" + concreteSingleIndex + "]/[" + item.type() + "]/[" + item.id() + "]";
+                responses.set(i, newItemFailure(concreteSingleIndex, item.type(), item.id(), new IllegalArgumentException(message)));
+                continue;
+            }
```

Comment:
I was looking at that code as well, but I don't think it will reach that line. Since the `aliasOrIndex` parameter can only be a concrete index the code should return on this line? https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java#L405

I can move that code into the try catch if we want to be defensive. Perhaps its better to move everything in that for loop into the try catch?

Another question: is it possible for an index to be deleted (or some other cluster change) while the code is inside that loop? For example when `indexNameExpressionResolver.concreteSingleIndex` is called the index exists, but when it gets down to `ShardId shardId =....` the index has been deleted?
Thanks